packages/pubsub/src/batch.js

```diff
@@ -0,0 +1,220 @@
+/*!
+ * Copyright 2014 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/batch
+ */
+
+'use strict';
+
+var prop = require('propprop');
+
+/**
+ * Semi-generic Batch object used to queue up multiple requests while limiting
+ * the number of concurrent requests.
+ *
+ * @constructor
+ * @alias module:pubsub/batch
+ *
+ * @param {object} options - Batch configuration settings.
+ * @param {number} options.maxBytes - Max size of data to be sent per request.
+ * @param {number} options.maxMessages - Max number of messages to be sent per
+ *     request.
+ * @param {number} options.maxMilliseconds - Max time to wait before sending
+ *     data.
+ * @param {number} options.maxRequests - Max number of concurrent requests
```

Comment:
maxRequests seems to undermine the other batch configuration settings. If a user specifies maxMilliseconds, for example, we must try send the batch after that time elapses. We can't wait for concurrent requests to complete.